File: algorithms/04_java/strings/BoyerMoore.java
public final class BoyerMoore {
    private BoyerMoore() {}

    public static int search(String text, String pattern) {
        int n = text.length(), m = pattern.length();
        int[] skip = new int[256];
        for (int i = 0; i < 256; i++) {
            skip[i] = m;
        }
        for (int i = 0; i < m - 1; i++) {
            skip[pattern.charAt(i)] = m - 1 - i;
        }
        int i = 0;
        while (i <= n - m) {
            int j = m - 1;
            while (j >= 0 && text.charAt(i + j) == pattern.charAt(j)) {
                j--;
            }
            if (j < 0) {
                return i;
            }
            i += skip[text.charAt(i + m - 1)];
        }
        return -1;
    }

    public static void main(String[] args) {
        if (search("here is a simple example", "example") != 17 || search("abc", "xyz") != -1) {
            throw new AssertionError("search failed");
        }
        System.out.println("boyer-moore ok");
    }
}

File: algorithms/07_typescript/strings/boyer_moore.ts
// Boyer-Moore-Horspool substring search.
function boyerMoore(text: string, pattern: string): number {
  const n = text.length;
  const m = pattern.length;
  const skip = new Array<number>(256).fill(m);
  for (let i = 0; i < m - 1; i += 1) skip[pattern.charCodeAt(i)] = m - 1 - i;
  let i = 0;
  while (i + m <= n) {
    let j = m - 1;
    while (j >= 0 && text[i + j] === pattern[j]) j -= 1;
    if (j < 0) return i;
    i += skip[text.charCodeAt(i + m - 1)];
  }
  return -1;
}

if (boyerMoore("here is a simple example", "example") !== 17 || boyerMoore("abc", "xyz") !== -1) {
  throw new Error("search failed");
}
console.log("boyer-moore ok");
